Sources/GRPC/Interceptor/ClientInterceptorPipeline.swift

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Logging
+import NIO
+import NIOHPACK
+import NIOHTTP2
+
+internal final class ClientInterceptorPipeline<Request, Response> {
+  /// A logger.
+  internal let logger: Logger
+
+  /// The `EventLoop` this RPC is being executed on.
+  internal let eventLoop: EventLoop
+
+  /// The contexts associated with the interceptors stored in this pipeline.
+  private let contexts: [ClientInterceptorContext<Request, Response>]
+
+  /// Returns the context for the given index, if one exists.
+  /// - Parameter index: The index of the `ClientInterceptorContext` to return.
+  /// - Returns: The `ClientInterceptorContext` or `nil` if one does not exist for the given index.
+  internal func context(atIndex index: Int) -> ClientInterceptorContext<Request, Response>? {
+    return self.contexts[checkingIndices: index]
+  }
+
+  /// The context closest to the `NIO.Channel`, i.e. where inbound events originate.
+  private var head: ClientInterceptorContext<Request, Response>? {
+    return self.contexts.first
+  }
+
+  /// The context closest to the application, i.e. where outbound events originate.
+  private var tail: ClientInterceptorContext<Request, Response>? {
+    return self.contexts.last
+  }
+
+  internal init(
+    eventLoop: EventLoop,
+    logger: Logger
+  ) {
+    self.eventLoop = eventLoop
+    self.logger = logger
+    self.contexts = []
```

Comment:
This doesn't seem useful given that `self.contexts` is `let`. Presumably we'll want to take them as an argument?

Reply:
Eventually, yes. This the pipeline is mostly just placeholder at the moment.